src/dsp/hub.ts

```diff
@@ -48,14 +57,18 @@ export class DSPHub {
   }
 
   toPrimaryKey(sigKeys: any[]) {
-    return sigKeys.map((keys: any[]) => keys.join('.')).join('-')
+    return sigKeys.map((keys: any[]) => keys.join('.')).join('-') as DSPHubPrimaryKey
   }
 
-  setDSP(primaryKey: string, foreignKeys: DSPHubForeignKeys, dsp: InstanceType<typeof DataSourcePlugin>) {
+  setDSP(primaryKey: DSPHubPrimaryKey, foreignKeys: DSPHubForeignKeys, dsp: InstanceType<typeof DataSourcePlugin<any, any, any, any>>) {
     this.hub.set(this.toHubKey(primaryKey, foreignKeys), dsp)
   }
 
-  getDSP(primaryKey: any, foreignKeys: any) {
+  getDSP(primaryKey: 'ethereum', foreignKeys: ValueReset<DSPHubForeignKeys, 'isLocal', false>): InstanceType<typeof EthereumDataSourcePlugin> | undefined
+  getDSP(primaryKey: 'ethereum', foreignKeys: ValueReset<DSPHubForeignKeys, 'isLocal', true>): InstanceType<typeof EthereumLocalDataSourcePlugin> | undefined
+  getDSP(primaryKey: 'ethereum-offchain.bytes', foreignKeys: ValueReset<DSPHubForeignKeys, 'isLocal', false>): InstanceType<typeof EthereumOffchainDSP> | undefined
+  getDSP(primaryKey: DSPHubPrimaryKey, foreignKeys: DSPHubForeignKeys): InstanceType<typeof DataSourcePlugin> | undefined
+  getDSP(primaryKey: DSPHubPrimaryKey, foreignKeys: DSPHubForeignKeys): DSPTypes | undefined {
```

Comment:
Is this function overloading a bit redundant? My goal is to deduce the respective types based on different DSP names.

Reply:
yeah i think this is a bit redundant. i feel the hub can just use instance as they value type, keep the register at 1 place, as simple as possible. 
plus, didn't see the benefit of introducing DSPTypes

Reply:
When a certain DSP class adds other member methods, this approach can deduce the accurate type.
Otherwise, the editor may consider the newly added method as non-existent.

By offering such a variety of types, our main goal is to provide developers with a better development experience.
However, these actions may also be unnecessary.